lib/helpers/node-matcher.js

```diff
@@ -0,0 +1,52 @@
+'use strict';
+
+// Source: jscodeshift/src/matchNode.js
+const hasOwn = Object.prototype.hasOwnProperty.call.bind(Object.prototype.hasOwnProperty);
+
+function isNode(value) {
+  return typeof value === 'object' && value;
+}
+
+function matchNode(haystack, needle) {
+  if (typeof needle === 'function') {
+    return needle(haystack);
+  }
+  if (isNode(needle) && isNode(haystack)) {
+    return Object.keys(needle).every(function (property) {
+      return hasOwn(haystack, property) && matchNode(haystack[property], needle[property]);
+    });
+  }
+  return haystack === needle;
+}
+
+// Source: jscodeshift/src/core.js
+function matchOne(path, filter) {
+  if (typeof path.get === 'function') {
+    path = path.get();
+  } else {
+    path = { value: path };
+  }
+  return matchNode(path.value, filter);
+}
+
+// Wrappers
+function nodeMatchesRef(testNode, refNode) {
+  return testNode && refNode && matchOne(testNode, refNode);
+}
+
+function nodeMatchesValidRef(testNode, refNodes) {
+  return testNode && refNodes && refNodes.some((refNode) => nodeMatchesRef(testNode, refNode));
```

Comment:
p4/5

I think we know for sure that `testNode` and `refNodes` are not falsey here, (due to the check in the beginning of `match` for `if (!testNode || !ref) { return; }`), so we can remove the extra defensiveness (and make this method a little bit easier to read).

```suggestion
  return refNodes.some((refNode) => nodeMatchesRef(testNode, refNode));
```

Reply:
p2/5

That suggestion got me thinking though, maybe we don't even need this method? That `refNodes.some((refNode) => nodeMatchesRef(testNode, refNode))` bit is probably simple enough to include inline in that `Array.isArray(ref)` branch of the `if` in `matches`.

What do you think?

Reply:
I absolutely agree with this approach for consolidating the existing behavior down to `matchOne` as the 'heavy lifting' match function and having `match` absorb both the input existence assertions and the 'check against an `Array` of `refNodes`' control/logic. 

They have been addressed in 8023b15 and a66c952, respectively. 👍

Reply:
Absolutely agree -- addressed both parts of that in a66c952 👍